File: encryption/atbash.py
from encryption.base import EncryptionMethod, alphabet

class AtbashEncryption(EncryptionMethod):
    def __init__(self):
        super().__init__()

    # reverses the char
    def __reverse_char(self, char: str) -> str:
        if char not in list(alphabet):
            return char
        
        return alphabet[abs((len(alphabet)-1) - list(alphabet).index(char))]
        
    # encrypts with atbash
    def encrypt(self, message: str) -> str:
        reversed_chars = [self.__reverse_char(char) for char in list(message.lower())]
        encrypted_result = ''.join(str(char) for char in reversed_chars)

        return encrypted_result

    # decrypting literally just takes the encrypted message and re-reverses it LOL
    def decrypt(self, cipher: str) -> str:
        return self.encrypt(cipher)

File: rust/src/encryption/atbash.rs
use super::{EncryptionMethod, grab_alphabet};

pub struct AtbashEncryption;

impl AtbashEncryption {
    fn reverse_char(&self, alphabet: &str, c: char) -> String {
        let alphabet_index = &alphabet.chars().position(|r| r == c);

        // if the character isn't in the alphabet, then add it to the string as itself
        match alphabet_index {
            Some(n) => {
                let reverse_c = alphabet.chars().nth((alphabet.len()-1) - *n);
                return reverse_c.unwrap().to_string();
            },
            None => c.to_string(),
        }
    }
}

impl EncryptionMethod for AtbashEncryption {
    fn encrypt(&self, message: &str) -> String {
        let mut encrypted_string = String::new();
        let alphabet = grab_alphabet();

        // loops through each character and reverses it
        for c in message.to_lowercase().chars() {
            let reversed_char = self.reverse_char(&alphabet, c);

            encrypted_string.push_str(&reversed_char);
        }

        encrypted_string
    }

    // literally just does the same thing LOL
    fn decrypt(&self, cipher: &str) -> String {
        self.encrypt(cipher)
    }
}